src/Core/Cart/Calculator.php

```diff
@@ -263,13 +263,15 @@ public function calculateRows()
 
     /**
      * calculate only cart rules (rows and fees have to be calculated first).
+     *
+     * @param bool $withFreeShipping used to calculate free shipping discount (avoid loop on shipping calculation)
      */
-    public function calculateCartRules()
+    public function calculateCartRules($withFreeShipping = true)
     {
         $this->cartRuleCalculator->setCartRules($this->cartRules)
             ->setCartRows($this->cartRows)
             ->setCalculator($this)
-            ->applyCartRules();
+            ->applyCartRules($withFreeShipping);
```

Comment:
You say you're not confident in the way you improved this. What about having properties for the free shipping?

```suggestion
            ->setFreeShipping($withFreeShipping);
            ->applyCartRules();
```

Reply:
good idea ;)